File: Frontend/trackr-administrador/src/app/shared/http/examen/clasificacion-pregunta.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { SimpleSelectorDto } from '@dtos/general/simple-selector-dto';
import { ClasificacionPreguntaFormularioDto } from '../../examen/clasificacion-pregunta-formulario-dto';
import { ClasificacionPreguntaGridDto } from '../../examen/clasificacion-pregunta-grid-dto';

@Injectable({
  providedIn: 'root'
})
export class ClasificacionPreguntaService {
  private dataUrl = 'clasificacionPregunta/';
  constructor(private http: HttpClient) {}

  public consultar(idClasificacionPregunta: number): Observable<ClasificacionPreguntaFormularioDto> {
    return this.http.get<ClasificacionPreguntaFormularioDto>(
      this.dataUrl + `${idClasificacionPregunta}`
    );
  }

  public consultarParaGrid(): Observable<ClasificacionPreguntaGridDto[]> {
    return this.http.get<ClasificacionPreguntaGridDto[]>(this.dataUrl);
  }

  public agregar(entidad:ClasificacionPreguntaGridDto): Observable<void> {
    return this.http.post<void>(this.dataUrl, entidad);
  }

  public editar(entidad: ClasificacionPreguntaGridDto): Observable<void> {
    return this.http.put<void>(this.dataUrl, entidad);
  }

  public eliminar(idClasificacionPregunta: number): Observable<void> {
    return this.http.delete<void>(this.dataUrl + `${idClasificacionPregunta}`);
  }

    public consultarSelector(): Observable<SimpleSelectorDto[]> {
        return this.http.get<SimpleSelectorDto[]>(this.dataUrl + `consultarSelector/`);
    }

}

File: Backend/TrackrAPI/TrackrAPI/Services/GestionExamen/ClasificacionPreguntaService.cs
using TrackrAPI.Dtos.General;
using TrackrAPI.Dtos.GestionExamen;
using TrackrAPI.Models;
using TrackrAPI.Repositorys.GestionExamen;

namespace TrackrAPI.Services.GestionExamen;

public class ClasificacionPreguntaService
{

    private readonly IClasificacionPreguntaRepository _clasificacionPreguntaRepository;

    public ClasificacionPreguntaService(IClasificacionPreguntaRepository clasificacionPreguntaRepository)
    {
        _clasificacionPreguntaRepository = clasificacionPreguntaRepository;
    }
    public IEnumerable<ClasificacionPreguntaGridDto> ConsultarParaGrid()
    {
        return _clasificacionPreguntaRepository.ConsultarParaGrid();
    }
    public ClasificacionPreguntaFormularioDto Consultar(int idClasificacionPregunta)
    {
        return _clasificacionPreguntaRepository.Consultar(idClasificacionPregunta);
    }
    public void Agregar(ClasificacionPreguntaFormularioDto dto)
    {
        try
        {
            var clasificacionPregunta = new ClasificacionPregunta
            {
                IdClasificacionPregunta = dto.IdClasificacionPregunta,
                Nombre = dto.Nombre,
                Estatus = true,
                Clave =  GenerarClave(),
            };
            _clasificacionPreguntaRepository.Agregar(clasificacionPregunta);

        }
        catch (Exception ex)
        {
            Console.WriteLine(ex.Message);
        }
    }

    private string GenerarClave()
    {
        var clasificacionPregunta = _clasificacionPreguntaRepository.ConsultarParaGrid().OrderByDescending(x => x.IdClasificacionPregunta).FirstOrDefault();

        return (clasificacionPregunta.IdClasificacionPregunta + 1).ToString();
    }

    public void Editar(ClasificacionPreguntaFormularioDto dto)
    {
            try
        {
            var clasificacionPregunta = new ClasificacionPregunta
            {
                IdClasificacionPregunta = dto.IdClasificacionPregunta,
                Nombre = dto.Nombre,
                Estatus = dto.Estatus,
                Clave =  dto.Clave,
            };
            _clasificacionPreguntaRepository.Editar(clasificacionPregunta);

        }
        catch (Exception ex)
        {
            Console.WriteLine(ex.Message);
        }
    }

    public void Eliminar(int idClasificacionPregunta)
    {
        _clasificacionPreguntaRepository.Eliminar(idClasificacionPregunta);
    }

    public IEnumerable<SimpleSelectorDto> ConsultarTodosParaSelector()
    {
        return _clasificacionPreguntaRepository.ConsultarTodosParaSelector();
    }
}
